ClassCastException on attempt to decorate list of page elements of generic type

## The problem
ClassCastException on attempt to decorate list of page elements of generic type

## Details
Let's say I have the following element container class:

```java
public class DummyTypedElement<T> extends ElementsContainer {

    @FindBy(tagName = "a")
    private List<T> list;

}
```

And a page object with field of that type:
```java
public class DummyPage {

    @FindBy(tagName = "div")
    private DummyTypedElement<SelenideElement> element;
}
```

When I tried to initialize that page by `Selenide.page` I got the following exception:
<pre>
Caused by: java.lang.RuntimeException: Failed to create elements container for field element
	at com.codeborne.selenide.impl.SelenideFieldDecorator.createElementsContainer(SelenideFieldDecorator.java:69) ~[selenide-4.9.1.jar:4.9.1]
	at com.codeborne.selenide.impl.SelenideFieldDecorator.decorate(SelenideFieldDecorator.java:38) ~[selenide-4.9.1.jar:4.9.1]
	at com.codeborne.selenide.SelenidePageFactory.proxyFields(SelenidePageFactory.java:35) ~[selenide-4.9.1.jar:4.9.1]
	at com.codeborne.selenide.SelenidePageFactory.initElements(SelenidePageFactory.java:24) ~[selenide-4.9.1.jar:4.9.1]
	at com.codeborne.selenide.Selenide.page(Selenide.java:643) ~[selenide-4.9.1.jar:4.9.1]
	at com.codeborne.selenide.Selenide.page(Selenide.java:632) ~[selenide-4.9.1.jar:4.9.1]
	... 60 more
Caused by: java.lang.ClassCastException: sun.reflect.generics.reflectiveObjects.TypeVariableImpl cannot be cast to java.lang.Class
	at com.codeborne.selenide.impl.SelenideFieldDecorator.getListGenericType(SelenideFieldDecorator.java:98) ~[selenide-4.9.1.jar:4.9.1]
	at com.codeborne.selenide.impl.SelenideFieldDecorator.isDecoratableList(SelenideFieldDecorator.java:88) ~[selenide-4.9.1.jar:4.9.1]
	at com.codeborne.selenide.impl.SelenideFieldDecorator.decorate(SelenideFieldDecorator.java:40) ~[selenide-4.9.1.jar:4.9.1]
	at com.codeborne.selenide.SelenidePageFactory.proxyFields(SelenidePageFactory.java:35) ~[selenide-4.9.1.jar:4.9.1]
	at com.codeborne.selenide.SelenidePageFactory.initElements(SelenidePageFactory.java:24) ~[selenide-4.9.1.jar:4.9.1]
	at com.codeborne.selenide.impl.SelenideFieldDecorator.initElementsContainer(SelenideFieldDecorator.java:78) ~[selenide-4.9.1.jar:4.9.1]
	at com.codeborne.selenide.impl.SelenideFieldDecorator.createElementsContainer(SelenideFieldDecorator.java:67) ~[selenide-4.9.1.jar:4.9.1]
	at com.codeborne.selenide.impl.SelenideFieldDecorator.decorate(SelenideFieldDecorator.java:38) ~[selenide-4.9.1.jar:4.9.1]
	at com.codeborne.selenide.SelenidePageFactory.proxyFields(SelenidePageFactory.java:35) ~[selenide-4.9.1.jar:4.9.1]
	at com.codeborne.selenide.SelenidePageFactory.initElements(SelenidePageFactory.java:24) ~[selenide-4.9.1.jar:4.9.1]
	at com.codeborne.selenide.Selenide.page(Selenide.java:643) ~[selenide-4.9.1.jar:4.9.1]
	at com.codeborne.selenide.Selenide.page(Selenide.java:632) ~[selenide-4.9.1.jar:4.9.1]
</pre>


## Tell us about your environment
* **Selenide Version:** 4.9.1
* **Chrome\Firefox\IE Version:**
* **Browser Driver Version:**
* **Selenium Version:**
* **OS Version:**

## Code To Reproduce Issue [ Good To Have ]
Please remember that with sample code it's easier to reproduce the bug.

--- a/src/main/java/com/codeborne/selenide/ElementsContainer.java
+++ b/src/main/java/com/codeborne/selenide/ElementsContainer.java
@@ -8,17 +8,14 @@ import javax.annotation.ParametersAreNonnullByDefault;
 public abstract class ElementsContainer {
   private SelenideElement self;
 
-  public ElementsContainer() {
-
-  }
-
+  /**
+   * @deprecated I rather think that this method is not needed.
+   * You are expected to find elements INSIDE this container, not the container itself.
+   */
   @CheckReturnValue
   @Nonnull
+  @Deprecated
   public SelenideElement getSelf() {
     return self;
-  }
-
-  public void setSelf(SelenideElement self) {
-    this.self = self;
   }
 }

--- a/src/main/java/com/codeborne/selenide/impl/SelenideFieldDecorator.java
+++ b/src/main/java/com/codeborne/selenide/impl/SelenideFieldDecorator.java
@@ -12,6 +12,8 @@ import org.openqa.selenium.support.FindBys;
 import org.openqa.selenium.support.pagefactory.Annotations;
 import org.openqa.selenium.support.pagefactory.DefaultElementLocatorFactory;
 import org.openqa.selenium.support.pagefactory.DefaultFieldDecorator;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
 
 import javax.annotation.CheckReturnValue;
 import javax.annotation.Nonnull;
@@ -19,14 +21,17 @@ import javax.annotation.Nullable;
 import javax.annotation.ParametersAreNonnullByDefault;
 import java.lang.reflect.Constructor;
 import java.lang.reflect.Field;
-import java.lang.reflect.InvocationTargetException;
+import java.lang.reflect.Modifier;
 import java.lang.reflect.ParameterizedType;
 import java.lang.reflect.Type;
+import java.lang.reflect.TypeVariable;
 import java.util.ArrayList;
+import java.util.Arrays;
 import java.util.List;
 
 @ParametersAreNonnullByDefault
 public class SelenideFieldDecorator extends DefaultFieldDecorator {
+  private static final Logger logger = LoggerFactory.getLogger(SelenideFieldDecorator.class);
   private final SelenidePageFactory pageFactory;
   private final Driver driver;
   private final SearchContext searchContext;
@@ -38,25 +43,38 @@ public class SelenideFieldDecorator extends DefaultFieldDecorator {
     this.searchContext = searchContext;
   }
 
-  @Override
   @CheckReturnValue
   @Nullable
-  public Object decorate(ClassLoader loader, Field field) {
+  @Override
+  public final Object decorate(ClassLoader loader, Field field) {
+    Type[] classGenericTypes = field.getDeclaringClass().getGenericInterfaces();
+    return decorate(loader, field, classGenericTypes);
+  }
+
+  @CheckReturnValue
+  @Nullable
+  public final Object decorate(ClassLoader loader, Field field, Type[] genericTypes) {
     if (ElementsContainer.class.equals(field.getDeclaringClass()) && "self".equals(field.getName())) {
-      return searchContext;
+      if (searchContext instanceof SelenideElement) {
+        return searchContext;
+      }
+      else {
+        logger.warn("Cannot initialize field {}", field);
+        return null;
+      }
     }
     By selector = new Annotations(field).buildBy();
     if (WebElement.class.isAssignableFrom(field.getType())) {
       return ElementFinder.wrap(driver, searchContext, selector, 0);
     }
-    if (ElementsCollection.class.isAssignableFrom(field.getType()) || isDecoratableList(field, WebElement.class)) {
+    if (ElementsCollection.class.isAssignableFrom(field.getType()) || isDecoratableList(field, genericTypes, WebElement.class)) {
       return new ElementsCollection(new BySelectorCollection(driver, searchContext, selector));
     }
     else if (ElementsContainer.class.isAssignableFrom(field.getType())) {
       return createElementsContainer(selector, field);
     }
-    else if (isDecoratableList(field, ElementsContainer.class)) {
-      return createElementsContainerList(field);
+    else if (isDecoratableList(field, genericTypes, ElementsContainer.class)) {
+      return createElementsContainerList(field, genericTypes);
     }
 
     return super.decorate(loader, field);
@@ -64,9 +82,9 @@ public class SelenideFieldDecorator extends DefaultFieldDecorator {
 
   @CheckReturnValue
   @Nonnull
-  private List<ElementsContainer> createElementsContainerList(Field field) {
+  private List<ElementsContainer> createElementsContainerList(Field field, Type[] genericTypes) {
     List<ElementsContainer> result = new ArrayList<>();
-    Class<?> listType = getListGenericType(field);
+    Class<?> listType = getListGenericType(field, genericTypes);
     if (listType == null) {
       throw new IllegalArgumentException("Cannot detect list type for " + field);
     }
@@ -74,10 +92,15 @@ public class SelenideFieldDecorator extends DefaultFieldDecorator {
     try {
       List<SelenideElement> selfList = SelenideElementListProxy.wrap(driver, factory.createLocator(field));
       for (SelenideElement element : selfList) {
-        result.add(initElementsContainer(listType, element));
+        Type[] types = ((ParameterizedType) field.getGenericType()).getActualTypeArguments();
+        result.add(initElementsContainer(field, element, listType, types));
       }
       return result;
-    } catch (Exception e) {
+    }
+    catch (RuntimeException e) {
+      throw e;
+    }
+    catch (Exception e) {
       throw new RuntimeException("Failed to create elements container list for field " + field.getName(), e);
     }
   }
@@ -87,41 +110,76 @@ public class SelenideFieldDecorator extends DefaultFieldDecorator {
   private ElementsContainer createElementsContainer(By selector, Field field) {
     try {
       SelenideElement self = ElementFinder.wrap(driver, searchContext, selector, 0);
-      return initElementsContainer(field.getType(), self);
-    } catch (Exception e) {
+      return initElementsContainer(field, self);
+    }
+    catch (RuntimeException e) {
+      throw e;
+    }
+    catch (Exception e) {
       throw new RuntimeException("Failed to create elements container for field " + field.getName(), e);
     }
   }
 
+  private ElementsContainer initElementsContainer(Field field, SelenideElement self) throws Exception {
+    Type[] genericTypes = field.getGenericType() instanceof ParameterizedType ?
+      ((ParameterizedType) field.getGenericType()).getActualTypeArguments() : new Type[0];
+    return initElementsContainer(field, self, field.getType(), genericTypes);
+  }
+
   @CheckReturnValue
   @Nonnull
-  private ElementsContainer initElementsContainer(Class<?> type, SelenideElement self)
-      throws InstantiationException, IllegalAccessException, NoSuchMethodException, InvocationTargetException {
+  private ElementsContainer initElementsContainer(Field field,
+                                                  SelenideElement self,
+                                                  Class<?> type,
+                                                  Type[] genericTypes) throws Exception {
+    if (Modifier.isInterface(type.getModifiers())) {
+      throw new IllegalArgumentException("Cannot initialize field " + field + ": " + type + " is interface");
+    }
+    if (Modifier.isAbstract(type.getModifiers())) {
+      throw new IllegalArgumentException("Cannot initialize field " + field + ": " + type + " is abstract");
+    }
     Constructor<?> constructor = type.getDeclaredConstructor();
     constructor.setAccessible(true);
     ElementsContainer result = (ElementsContainer) constructor.newInstance();
-    pageFactory.initElements(new SelenideFieldDecorator(pageFactory, driver, self), result);
+    pageFactory.initElements(new SelenideFieldDecorator(pageFactory, driver, self), result, genericTypes);
     return result;
   }
 
   @CheckReturnValue
-  private boolean isDecoratableList(Field field, Class<?> type) {
+  private boolean isDecoratableList(Field field, Type[] genericTypes, Class<?> type) {
     if (!List.class.isAssignableFrom(field.getType())) {
       return false;
     }
 
-    Class<?> listType = getListGenericType(field);
+    Class<?> listType = getListGenericType(field, genericTypes);
 
     return listType != null && type.isAssignableFrom(listType)
-        && (field.getAnnotation(FindBy.class) != null || field.getAnnotation(FindBys.class) != null);
+      && (field.getAnnotation(FindBy.class) != null || field.getAnnotation(FindBys.class) != null);
   }
 
   @CheckReturnValue
   @Nullable
-  private Class<?> getListGenericType(Field field) {
-    Type genericType = field.getGenericType();
-    if (!(genericType instanceof ParameterizedType)) return null;
+  private Class<?> getListGenericType(Field field, Type[] genericTypes) {
+    Type fieldType = field.getGenericType();
+    if (!(fieldType instanceof ParameterizedType)) return null;
 
-    return (Class<?>) ((ParameterizedType) genericType).getActualTypeArguments()[0];
+    Type[] actualTypeArguments = ((ParameterizedType) fieldType).getActualTypeArguments();
+    Type firstType = actualTypeArguments[0];
+    if (firstType instanceof TypeVariable) {
+      int indexOfType = indexOf(field.getDeclaringClass(), firstType);
+      return (Class<?>) genericTypes[indexOfType];
+    }
+    else if (firstType instanceof Class) {
+      return (Class<?>) firstType;
+    }
+    throw new IllegalArgumentException("Cannot detect list type of " + field);
+  }
+
+  private int indexOf(Class<?> klass, Type firstArgument) {
+    Object[] objects = Arrays.stream(klass.getTypeParameters()).toArray();
+    for (int i = 0; i < objects.length; i++) {
+      if (objects[i].equals(firstArgument)) return i;
+    }
+    return -1;
   }
 }

--- a/src/main/java/com/codeborne/selenide/impl/SelenideFieldDecorator.java
+++ b/src/main/java/com/codeborne/selenide/impl/SelenideFieldDecorator.java
@@ -46,7 +46,7 @@ public class SelenideFieldDecorator extends DefaultFieldDecorator {
     if (WebElement.class.isAssignableFrom(field.getType())) {
       return ElementFinder.wrap(driver, searchContext, selector, 0);
     }
-    if (ElementsCollection.class.isAssignableFrom(field.getType())) {
+    if (ElementsCollection.class.isAssignableFrom(field.getType()) || isDecoratableList(field, SelenideElement.class)) {
       return new ElementsCollection(new BySelectorCollection(driver, searchContext, selector));
     }
     else if (ElementsContainer.class.isAssignableFrom(field.getType())) {
@@ -54,9 +54,6 @@ public class SelenideFieldDecorator extends DefaultFieldDecorator {
     }
     else if (isDecoratableList(field, ElementsContainer.class)) {
       return createElementsContainerList(field);
-    }
-    else if (isDecoratableList(field, SelenideElement.class)) {
-      return SelenideElementListProxy.wrap(driver, factory.createLocator(field));
     }
 
     return super.decorate(loader, field);

--- a/src/main/java/com/codeborne/selenide/impl/SelenideFieldDecorator.java
+++ b/src/main/java/com/codeborne/selenide/impl/SelenideFieldDecorator.java
@@ -46,7 +46,7 @@ public class SelenideFieldDecorator extends DefaultFieldDecorator {
     if (WebElement.class.isAssignableFrom(field.getType())) {
       return ElementFinder.wrap(driver, searchContext, selector, 0);
     }
-    if (ElementsCollection.class.isAssignableFrom(field.getType()) || isDecoratableList(field, SelenideElement.class)) {
+    if (ElementsCollection.class.isAssignableFrom(field.getType()) || isDecoratableList(field, WebElement.class)) {
       return new ElementsCollection(new BySelectorCollection(driver, searchContext, selector));
     }
     else if (ElementsContainer.class.isAssignableFrom(field.getType())) {

--- a/src/main/java/com/codeborne/selenide/impl/SelenideFieldDecorator.java
+++ b/src/main/java/com/codeborne/selenide/impl/SelenideFieldDecorator.java
@@ -42,6 +42,9 @@ public class SelenideFieldDecorator extends DefaultFieldDecorator {
   @CheckReturnValue
   @Nullable
   public Object decorate(ClassLoader loader, Field field) {
+    if (ElementsContainer.class.equals(field.getDeclaringClass()) && "self".equals(field.getName())) {
+      return searchContext;
+    }
     By selector = new Annotations(field).buildBy();
     if (WebElement.class.isAssignableFrom(field.getType())) {
       return ElementFinder.wrap(driver, searchContext, selector, 0);
@@ -98,7 +101,6 @@ public class SelenideFieldDecorator extends DefaultFieldDecorator {
     constructor.setAccessible(true);
     ElementsContainer result = (ElementsContainer) constructor.newInstance();
     pageFactory.initElements(new SelenideFieldDecorator(pageFactory, driver, self), result);
-    result.setSelf(self);
     return result;
   }
 

--- a/src/main/java/com/codeborne/selenide/impl/SelenidePageFactory.java
+++ b/src/main/java/com/codeborne/selenide/impl/SelenidePageFactory.java
@@ -8,6 +8,7 @@ import javax.annotation.Nonnull;
 import javax.annotation.ParametersAreNonnullByDefault;
 import java.lang.reflect.Constructor;
 import java.lang.reflect.Field;
+import java.lang.reflect.Type;
 
 /**
  * Factory class to make using Page Objects simpler and easier.
@@ -32,7 +33,8 @@ public class SelenidePageFactory {
   @CheckReturnValue
   @Nonnull
   public <PageObjectClass, T extends PageObjectClass> PageObjectClass page(Driver driver, T pageObject) {
-    initElements(new SelenideFieldDecorator(this, driver, driver.getWebDriver()), pageObject);
+    Type[] types = pageObject.getClass().getGenericInterfaces();
+    initElements(new SelenideFieldDecorator(this, driver, driver.getWebDriver()), pageObject, types);
     return pageObject;
   }
 
@@ -43,21 +45,21 @@ public class SelenidePageFactory {
    * @param decorator the decorator to use
    * @param page      The object to decorate the fields of
    */
-  public void initElements(FieldDecorator decorator, Object page) {
+  public void initElements(SelenideFieldDecorator decorator, Object page, Type[] genericTypes) {
     Class<?> proxyIn = page.getClass();
     while (proxyIn != Object.class) {
-      proxyFields(decorator, page, proxyIn);
+      proxyFields(decorator, page, proxyIn, genericTypes);
       proxyIn = proxyIn.getSuperclass();
     }
   }
 
-  private void proxyFields(FieldDecorator decorator, Object page, Class<?> proxyIn) {
+  private void proxyFields(SelenideFieldDecorator decorator, Object page, Class<?> proxyIn, Type[] genericTypes) {
     Field[] fields = proxyIn.getDeclaredFields();
     for (Field field : fields) {
       if (isInitialized(page, field)) {
         continue;
       }
-      Object value = decorator.decorate(page.getClass().getClassLoader(), field);
+      Object value = decorator.decorate(page.getClass().getClassLoader(), field, genericTypes);
       if (value != null) {
         try {
           field.setAccessible(true);

--- a/src/test/java/com/codeborne/selenide/impl/SelenideFieldDecoratorTest.java
+++ b/src/test/java/com/codeborne/selenide/impl/SelenideFieldDecoratorTest.java
@@ -88,8 +88,9 @@ final class SelenideFieldDecoratorTest implements WithAssertions {
   @Test
   void decoratesVanillaWebElements() throws NoSuchFieldException {
     final Object someDiv = fieldDecorator.decorate(getClass().getClassLoader(), getField("someDiv"));
+    assertThat(someDiv).isNotNull();
     assertThat(someDiv)
-      .withFailMessage("someDiv should not be instance of SelenideElement. Actual class: " + someDiv.getClass())
+      .withFailMessage("someDiv should be instance of SelenideElement. Actual class: " + someDiv.getClass())
       .isInstanceOf(SelenideElement.class);
   }
 

--- a/src/test/java/com/codeborne/selenide/impl/SelenideFieldDecoratorTest.java
+++ b/src/test/java/com/codeborne/selenide/impl/SelenideFieldDecoratorTest.java
@@ -3,7 +3,9 @@ package com.codeborne.selenide.impl;
 import com.codeborne.selenide.Config;
 import com.codeborne.selenide.Driver;
 import com.codeborne.selenide.DriverStub;
+import com.codeborne.selenide.ElementsCollection;
 import com.codeborne.selenide.ElementsContainer;
+import com.codeborne.selenide.SelenideConfig;
 import com.codeborne.selenide.SelenideElement;
 import org.assertj.core.api.WithAssertions;
 import org.junit.jupiter.api.Test;
@@ -20,11 +22,12 @@ import static java.util.Arrays.asList;
 import static org.mockito.Mockito.any;
 import static org.mockito.Mockito.mock;
 import static org.mockito.Mockito.verify;
+import static org.mockito.Mockito.verifyNoMoreInteractions;
 import static org.mockito.Mockito.when;
 
 final class SelenideFieldDecoratorTest implements WithAssertions {
   private final TestPage page = new TestPage();
-  private final Config config = mock(Config.class);
+  private final Config config = new SelenideConfig();
   private final WebDriver webDriver = mock(WebDriver.class);
   private final Driver driver = new DriverStub(config, null, webDriver, null);
   private final SelenidePageFactory pageFactory = new SelenidePageFactory();
@@ -45,26 +48,39 @@ final class SelenideFieldDecoratorTest implements WithAssertions {
   }
 
   private Field getField(String fieldName) throws NoSuchFieldException {
+    return getField(page, fieldName);
+  }
+
+  private <T> Field getField(T page, String fieldName) throws NoSuchFieldException {
     return page.getClass().getDeclaredField(fieldName);
   }
 
   @Test
-  @SuppressWarnings("unchecked")
+  void decoratesElementsCollection() throws NoSuchFieldException {
+    TestPageWithElementsCollection page = new TestPageWithElementsCollection();
+
+    Object decoratedField = fieldDecorator.decorate(getClass().getClassLoader(), getField(page, "rows"));
+
+    assertThat(decoratedField).isInstanceOf(ElementsCollection.class);
+    verifyNoMoreInteractions(webDriver);
+  }
+
+  @Test
   void decoratesListOfSelenideElements() throws NoSuchFieldException {
-    when(webDriver.findElements(any(By.class))).thenReturn(asList(mock(WebElement.class), mock(WebElement.class)));
+    WebElement element1 = mock(WebElement.class);
+    WebElement element2 = mock(WebElement.class);
+    when(webDriver.findElements(any(By.class))).thenReturn(asList(element1, element2));
 
     Object decoratedField = fieldDecorator.decorate(getClass().getClassLoader(), getField("rows"));
+    assertThat(decoratedField).isInstanceOf(ElementsCollection.class);
+    verifyNoMoreInteractions(webDriver);
 
-    assertThat(decoratedField)
-      .isInstanceOf(List.class);
-    List<SelenideElement> elements = (List<SelenideElement>) decoratedField;
-    assertThat(elements)
-      .hasSize(2);
-    verify(webDriver).findElements(any(By.class));
-    assertThat(elements.get(0))
-      .isInstanceOf(SelenideElement.class);
-    assertThat(elements.get(1))
-      .isInstanceOf(SelenideElement.class);
+    ElementsCollection elementsCollection = (ElementsCollection) decoratedField;
+    Object[] elements = elementsCollection.toArray();
+    assertThat(elements).hasSize(2);
+    verify(webDriver).findElements(By.cssSelector("table tbody tr"));
+    assertThat(elements[0]).isInstanceOf(SelenideElement.class);
+    assertThat(elements[0]).isInstanceOf(SelenideElement.class);
   }
 
   @Test
@@ -78,15 +94,16 @@ final class SelenideFieldDecoratorTest implements WithAssertions {
   @Test
   @SuppressWarnings("unchecked")
   void decoratesListOfVanillaWebElements() throws NoSuchFieldException {
-    when(webDriver.findElements(any(By.class))).thenReturn(asList(mock(WebElement.class), mock(WebElement.class)));
+    WebElement element1 = mock(WebElement.class);
+    WebElement element2 = mock(WebElement.class);
+    when(webDriver.findElements(any(By.class))).thenReturn(asList(element1, element2));
+
     List<WebElement> elements = (List<WebElement>) fieldDecorator.decorate(getClass().getClassLoader(), getField("data"));
-    assertThat(elements)
-      .hasSize(2);
-    verify(webDriver).findElements(any(By.class));
-    assertThat(elements.get(0))
-      .isInstanceOf(WebElement.class);
-    assertThat(elements.get(1))
-      .isNotInstanceOf(SelenideElement.class);
+
+    assertThat(elements).hasSize(2);
+    verify(webDriver).findElements(By.cssSelector("table tbody tr"));
+    assertThat(elements.get(0)).isInstanceOf(SelenideElement.class);
+    assertThat(elements.get(1)).isInstanceOf(SelenideElement.class);
   }
 
   @Test
@@ -177,5 +194,10 @@ final class SelenideFieldDecoratorTest implements WithAssertions {
 
     @FindBy(className = "name")
     SelenideElement name;
+  }
+
+  static class TestPageWithElementsCollection {
+    @FindBy(css = "table tbody tr")
+    ElementsCollection rows;
   }
 }

--- a/src/test/java/com/codeborne/selenide/impl/SelenideFieldDecoratorTest.java
+++ b/src/test/java/com/codeborne/selenide/impl/SelenideFieldDecoratorTest.java
@@ -18,6 +18,8 @@ import org.openqa.selenium.support.pagefactory.DefaultElementLocatorFactory;
 import java.lang.reflect.Field;
 import java.util.List;
 
+import static com.codeborne.selenide.Condition.text;
+import static com.codeborne.selenide.Mocks.mockWebElement;
 import static java.util.Arrays.asList;
 import static org.mockito.Mockito.any;
 import static org.mockito.Mockito.mock;
@@ -115,24 +117,24 @@ final class SelenideFieldDecoratorTest implements WithAssertions {
   @Test
   void decoratesElementsContainerWithItsSubElements() throws NoSuchFieldException {
     StatusBlock status = (StatusBlock) fieldDecorator.decorate(getClass().getClassLoader(), getField("status"));
-    WebElement statusElement = mock(WebElement.class);
+    WebElement statusElement = mockWebElement("div", "the status");
+    WebElement lastLogin = mockWebElement("div", "03.03.2003");
+    WebElement name = mockWebElement("div", "lena");
     when(webDriver.findElement(By.id("status"))).thenReturn(statusElement);
-    when(statusElement.findElement(By.className("last-login"))).thenReturn(mock(WebElement.class));
-    when(statusElement.findElement(By.className("name"))).thenReturn(mock(WebElement.class));
+    when(statusElement.findElement(By.className("last-login"))).thenReturn(lastLogin);
+    when(statusElement.findElement(By.className("name"))).thenReturn(name);
 
-    assertThat(status)
-      .isNotNull();
-    assertThat(status.getSelf())
-      .isNotNull();
-    status.getSelf().getText();
+    assertThat(status).isNotNull();
+    assertThat(status.getSelf()).isNotNull();
+    status.getSelf().shouldHave(text("the status"));
     verify(webDriver).findElement(By.id("status"));
-    assertThat(status.lastLogin)
-      .isNotNull();
-    status.lastLogin.getText();
+
+    assertThat(status.lastLogin).isNotNull();
+    status.lastLogin.shouldHave(text("03.03.2003"));
     verify(statusElement).findElement(By.className("last-login"));
-    assertThat(status.name)
-      .isNotNull();
-    status.name.getText();
+
+    assertThat(status.name).isNotNull();
+    status.name.shouldHave(text("lena"));
     verify(statusElement).findElement(By.className("name"));
   }
 
@@ -143,30 +145,28 @@ final class SelenideFieldDecoratorTest implements WithAssertions {
     WebElement statusElement2 = mock(WebElement.class);
     when(webDriver.findElements(any(By.class))).thenReturn(asList(statusElement1, statusElement2));
     when(statusElement1.getText()).thenReturn("status element1 text");
-    when(statusElement1.findElement(By.className("last-login"))).thenReturn(mock(WebElement.class));
-    when(statusElement1.findElement(By.className("name"))).thenReturn(mock(WebElement.class));
-    when(statusElement2.findElement(By.className("last-login"))).thenReturn(mock(WebElement.class));
-    when(statusElement2.findElement(By.className("name"))).thenReturn(mock(WebElement.class));
+    WebElement lastLogin1 = mockWebElement("div", "01.01.2001");
+    when(statusElement1.findElement(By.className("last-login"))).thenReturn(lastLogin1);
+    WebElement name1 = mockWebElement("div", "john");
+    when(statusElement1.findElement(By.className("name"))).thenReturn(name1);
+    WebElement lastLogin2 = mockWebElement("div", "02.02.2002");
+    when(statusElement2.findElement(By.className("last-login"))).thenReturn(lastLogin2);
+    WebElement name2 = mockWebElement("div", "katie");
+    when(statusElement2.findElement(By.className("name"))).thenReturn(name2);
 
     Object decoratedField = fieldDecorator.decorate(getClass().getClassLoader(), getField("statusHistory"));
 
-    assertThat(decoratedField)
-      .isInstanceOf(List.class);
+    assertThat(decoratedField).isInstanceOf(List.class);
     List<StatusBlock> statusHistory = (List<StatusBlock>) decoratedField;
-    assertThat(statusHistory)
-      .isNotNull();
+    assertThat(statusHistory).isNotNull();
     verify(webDriver).findElements(By.cssSelector("table.history tr.status"));
-    assertThat(statusHistory)
-      .hasSize(2);
-    assertThat(statusHistory.get(0).getSelf().getText())
-      .isEqualTo("status element1 text");
-    assertThat(statusHistory.get(0).lastLogin)
-      .isNotNull();
-    statusHistory.get(0).lastLogin.getText();
+    assertThat(statusHistory).hasSize(2);
+    assertThat(statusHistory.get(0).getSelf().getText()).isEqualTo("status element1 text");
+    assertThat(statusHistory.get(0).lastLogin).isNotNull();
+    statusHistory.get(0).lastLogin.shouldHave(text("01.01.2001"));
     verify(statusElement1).findElement(By.className("last-login"));
-    assertThat(statusHistory.get(0).name)
-      .isNotNull();
-    statusHistory.get(0).name.getText();
+    assertThat(statusHistory.get(0).name).isNotNull();
+    statusHistory.get(0).name.shouldHave(text("john"));
     verify(statusElement1).findElement(By.className("name"));
   }
 

--- a/statics/src/test/java/integration/pageobjects/DynamicListOfSelenideElementsTest.java
+++ b/statics/src/test/java/integration/pageobjects/DynamicListOfSelenideElementsTest.java
@@ -1,0 +1,47 @@
+package integration.pageobjects;
+
+import com.codeborne.selenide.ElementsContainer;
+import com.codeborne.selenide.SelenideElement;
+import com.codeborne.selenide.ex.ElementNotFound;
+import integration.IntegrationTest;
+import org.junit.jupiter.api.BeforeEach;
+import org.junit.jupiter.api.Test;
+import org.openqa.selenium.support.FindBy;
+
+import java.util.List;
+
+import static com.codeborne.selenide.Condition.text;
+import static com.codeborne.selenide.Configuration.timeout;
+import static com.codeborne.selenide.Selenide.page;
+import static org.assertj.core.api.Assertions.assertThatThrownBy;
+
+public class DynamicListOfSelenideElementsTest extends IntegrationTest {
+  @BeforeEach
+  void openTestPage() {
+    timeout = 1000;
+    openFile("page_with_selects_without_jquery.html");
+  }
+
+  @Test
+  void shouldWaitUntilCollectionIsLoaded() {
+    MyPage page = page(MyPage.class);
+    page.container.spans.get(0).shouldHave(text("dynamic content"));
+    page.container.spans.get(1).shouldHave(text("dynamic content2"));
+
+    assertThatThrownBy(() ->
+      page.container.spans.get(2).shouldHave(text("enough."))
+    )
+      .isInstanceOf(ElementNotFound.class)
+      .hasMessageStartingWith("Element not found {#dynamic-content-container/span[2]}");
+  }
+
+  private static class MyPage {
+    @FindBy(css = "#dynamic-content-container")
+    DynamicContentContainer container;
+  }
+
+  private static class DynamicContentContainer extends ElementsContainer {
+    @FindBy(css = "span")
+    List<SelenideElement> spans;
+  }
+}

--- a/statics/src/test/java/integration/pageobjects/DynamicListOfWebElementsTest.java
+++ b/statics/src/test/java/integration/pageobjects/DynamicListOfWebElementsTest.java
@@ -1,0 +1,50 @@
+package integration.pageobjects;
+
+import com.codeborne.selenide.ElementsContainer;
+import com.codeborne.selenide.ex.ElementNotFound;
+import integration.IntegrationTest;
+import org.junit.jupiter.api.BeforeEach;
+import org.junit.jupiter.api.Test;
+import org.openqa.selenium.WebElement;
+import org.openqa.selenium.support.FindBy;
+
+import java.util.List;
+
+import static com.codeborne.selenide.Condition.text;
+import static com.codeborne.selenide.Configuration.timeout;
+import static com.codeborne.selenide.Selenide.$;
+import static com.codeborne.selenide.Selenide.page;
+import static com.codeborne.selenide.Selenide.sleep;
+import static org.assertj.core.api.Assertions.assertThatThrownBy;
+
+public class DynamicListOfWebElementsTest extends IntegrationTest {
+  @BeforeEach
+  void openTestPage() {
+    timeout = 1000;
+    openFile("page_with_selects_without_jquery.html");
+  }
+
+  @Test
+  void shouldWaitUntilCollectionIsLoaded() {
+    sleep(500);
+    MyPage page = page(MyPage.class);
+    $(page.container.spans.get(0)).shouldHave(text("dynamic content"));
+    $(page.container.spans.get(1)).shouldHave(text("dynamic content2"));
+
+    assertThatThrownBy(() ->
+      $(page.container.spans.get(2)).shouldHave(text("enough."))
+    )
+      .isInstanceOf(ElementNotFound.class)
+      .hasMessageStartingWith("Element not found {#dynamic-content-container/span[2]}");
+  }
+
+  private static class MyPage {
+    @FindBy(css = "#dynamic-content-container")
+    DynamicContentContainer container;
+  }
+
+  private static class DynamicContentContainer extends ElementsContainer {
+    @FindBy(css = "span")
+    List<WebElement> spans;
+  }
+}

--- a/statics/src/test/java/integration/pageobjects/ElementsContainerWithManuallyInitializedFieldsTest.java
+++ b/statics/src/test/java/integration/pageobjects/ElementsContainerWithManuallyInitializedFieldsTest.java
@@ -1,8 +1,9 @@
-package integration;
+package integration.pageobjects;
 
 import com.codeborne.selenide.ElementsCollection;
 import com.codeborne.selenide.ElementsContainer;
 import com.codeborne.selenide.SelenideElement;
+import integration.IntegrationTest;
 import org.junit.jupiter.api.BeforeEach;
 import org.junit.jupiter.api.Test;
 import org.openqa.selenium.support.FindBy;

--- a/statics/src/test/java/integration/pageobjects/FieldOfGenericTypeTest.java
+++ b/statics/src/test/java/integration/pageobjects/FieldOfGenericTypeTest.java
@@ -1,0 +1,80 @@
+package integration.pageobjects;
+
+import com.codeborne.selenide.ElementsContainer;
+import com.codeborne.selenide.SelenideElement;
+import integration.IntegrationTest;
+import org.junit.jupiter.api.BeforeEach;
+import org.junit.jupiter.api.Test;
+import org.openqa.selenium.WebElement;
+import org.openqa.selenium.support.FindBy;
+
+import java.util.List;
+
+import static com.codeborne.selenide.Selenide.$;
+import static com.codeborne.selenide.Selenide.page;
+import static org.assertj.core.api.Assertions.assertThat;
+import static org.assertj.core.api.Assertions.assertThatThrownBy;
+
+public class FieldOfGenericTypeTest extends IntegrationTest {
+  @BeforeEach
+  void openTestPage() {
+    openFile("page_with_selects_without_jquery.html");
+  }
+
+  @Test
+  void injectsFoundSelenideElementAsSelf() {
+    DummyPage page = page(DummyPage.class);
+    assertThat(page).isInstanceOf(DummyPage.class);
+    assertThat(page.body).isInstanceOf(DummyTypedElement.class);
+    assertThat(page.body.selects).isInstanceOf(List.class);
+    assertThat(page.body.selects).hasSize(3);
+    assertThat(page.body.getSelf()).isEqualTo($("body"));
+    assertThat(page.body.selects.get(0)).isEqualTo($("select[name=domain]"));
+    assertThat(page.body.selects.get(1)).isEqualTo($("select#hero"));
+    assertThat(page.body.selects.get(2)).isEqualTo($("select#gender"));
+  }
+
+  @Test
+  void injectsFoundSelenideElementAsSelf2() {
+    AnotherPage page = page(AnotherPage.class);
+    assertThat(page.body.selects).hasSize(3);
+    assertThat(page.body).isInstanceOf(ElementsContainer.class);
+    assertThat(page.body.getSelf()).isEqualTo($("body"));
+    assertThat(page.body.selects.get(0)).isInstanceOf(WebElement.class);
+    assertThat(page.body.selects.get(0)).isEqualTo($("select[name=domain]"));
+    assertThat(page.body.selects.get(1)).isEqualTo($("select#hero"));
+    assertThat(page.body.selects.get(2)).isEqualTo($("select#gender"));
+  }
+
+  @Test
+  void injectsFoundSelenideElementAsSelf3() {
+    assertThatThrownBy(() -> page(YetAnotherPage.class))
+      .isInstanceOf(RuntimeException.class)
+      .hasMessageMatching("Failed to create new instance of class .*\\.FieldOfGenericTypeTest.YetAnotherPage.*")
+      .getCause()
+      .isInstanceOf(IllegalArgumentException.class)
+      .hasMessageMatching("Cannot initialize field .*List .*DummyTypedElement.selects: .*ElementsContainer is abstract");
+  }
+
+  static class DummyTypedElement<S, T> extends ElementsContainer {
+    private List<S> names;
+
+    @FindBy(tagName = "select")
+    private List<T> selects;
+  }
+
+  static class DummyPage {
+    @FindBy(tagName = "body")
+    private DummyTypedElement<String, SelenideElement> body;
+  }
+
+  static class AnotherPage {
+    @FindBy(tagName = "body")
+    private DummyTypedElement<Integer, WebElement> body;
+  }
+
+  static class YetAnotherPage {
+    @FindBy(tagName = "body")
+    private DummyTypedElement<Boolean, ElementsContainer> body;
+  }
+}

--- a/statics/src/test/java/integration/pageobjects/PageObjectTest.java
+++ b/statics/src/test/java/integration/pageobjects/PageObjectTest.java
@@ -1,10 +1,11 @@
-package integration;
+package integration.pageobjects;
 
 import com.codeborne.selenide.Condition;
 import com.codeborne.selenide.ElementsCollection;
 import com.codeborne.selenide.ElementsContainer;
 import com.codeborne.selenide.SelenideElement;
 import com.codeborne.selenide.ex.ElementNotFound;
+import integration.IntegrationTest;
 import org.junit.jupiter.api.BeforeEach;
 import org.junit.jupiter.api.Test;
 import org.openqa.selenium.WebElement;

--- a/statics/src/test/java/integration/pageobjects/PageObjectWithManuallyInitializedFieldsTest.java
+++ b/statics/src/test/java/integration/pageobjects/PageObjectWithManuallyInitializedFieldsTest.java
@@ -1,7 +1,8 @@
-package integration;
+package integration.pageobjects;
 
 import com.codeborne.selenide.Condition;
 import com.codeborne.selenide.SelenideElement;
+import integration.IntegrationTest;
 import org.junit.jupiter.api.BeforeEach;
 import org.junit.jupiter.api.Test;
 

--- a/statics/src/test/java/integration/pageobjects/SelenidePageFactoryTest.java
+++ b/statics/src/test/java/integration/pageobjects/SelenidePageFactoryTest.java
@@ -1,6 +1,7 @@
-package integration;
+package integration.pageobjects;
 
 import com.codeborne.selenide.SelenideElement;
+import integration.IntegrationTest;
 import org.junit.jupiter.api.BeforeEach;
 import org.junit.jupiter.api.Test;
 import org.openqa.selenium.support.FindBy;

--- a/statics/src/test/resources/fortest.html
+++ b/statics/src/test/resources/fortest.html
@@ -1,0 +1,14 @@
+<!DOCTYPE html>
+<html lang="en">
+<head>
+  <meta charset="UTF-8">
+  <title>Insert title here</title>
+</head>
+<body>
+<div class="content">
+  <form>
+    <h1>The header</h1>
+  </form>
+</div>
+</body>
+</html>

